### PR TITLE
Btfs 793 swarm connect

### PR DESCRIPTION
**Pull request by ShellyWEI**
Notice, the work flow can take longer because it need to find peer in network.

### CODE REVIEW

File: core/commands/storage.go [hunk 1]
```diff
@@ -438,12 +448,78 @@ func getValidHost(retryQueue *storage.RetryQueue) (*storage.HostNode, error) {
 				return nil, err
 			}
 		} else {
+			// find peer
+			pi, err := findPeer(ctx, n, nextHost.Identity)
+			if err != nil {
+				// cannot find peer, remove from queue permanently
+				log.Error(err)
```
Comment (taiyangc) [inline core/commands/storage.go]: It's common not able to find peer, shouldn't need so many logs

File: core/commands/storage.go [hunk 2]
```diff
@@ -438,12 +448,78 @@ func getValidHost(retryQueue *storage.RetryQueue) (*storage.HostNode, error) {
...
+				continue
+			}
+			if err := api.Swarm().Connect(ctx, *pi); err != nil {
+				log.Error("fail swarm connect first time")
```
Comment (taiyangc) [inline core/commands/storage.go]: too many logs

File: core/commands/storage.go [hunk 3]
```diff
@@ -438,12 +448,78 @@ func getValidHost(retryQueue *storage.RetryQueue) (*storage.HostNode, error) {
...
+				// force connect again
+				if err := api.Swarm().Connect(ctx, *pi); err != nil {
+					log.Error("fail swarm connect second time")
+					// in test mode, change the ip to 0.0.0.0
+					if test {
+						if err := changeAddress(pi); err != nil {
+							log.Error(err)
```
Comment (taiyangc) [inline core/commands/storage.go]: on err, you should skip `Connect` and go to the IncrementFail directly. Try to organize the fail path a little more.

for example (rough idea)

```
var err error
if err = api.Swarm.Connect(...); err != nil {
    <change address>
    if err == nil {
        err = api.Swarm.Connect(...)
    }
}
if err != nil {
   <increment fail>
   <offer>
   if err != nil { return ... }
   continue
}
```

File: core/commands/storage.go [hunk 4]
```diff
@@ -261,6 +268,7 @@ func controlSessionTimeout(ss *storage.Session) {
 				curStatus = sessionState.CurrentStep + 1
 				ss.SetStatus(curStatus)
 			} else {
+				log.Error(sessionState.Err)
```
[the remark this answers is not in the document]
Comment (ShellyWEI) [reply core/commands/storage.go]: Sure, mark it in code comment

File: core/commands/storage.go (hunk 1 above)
Comment (ShellyWEI) [reply core/commands/storage.go]: if not find peer, I would offer back to queue and increase fail times.

File: core/commands/storage.go (hunk 3 above)
Comment (ShellyWEI) [reply core/commands/storage.go]: So you mean give local connect also 2 times connect?

**Pull request merged**